src/NLog/Targets/TargetWithContext.cs

```diff
@@ -0,0 +1,802 @@
+﻿// 
+// Copyright (c) 2004-2017 Jaroslaw Kowalski <jaak@jkowalski.net>, Kim Christensen, Julian Verdurmen
+// 
+// All rights reserved.
+// 
+// Redistribution and use in source and binary forms, with or without 
+// modification, are permitted provided that the following conditions 
+// are met:
+// 
+// * Redistributions of source code must retain the above copyright notice, 
+//   this list of conditions and the following disclaimer. 
+// 
+// * Redistributions in binary form must reproduce the above copyright notice,
+//   this list of conditions and the following disclaimer in the documentation
+//   and/or other materials provided with the distribution. 
+// 
+// * Neither the name of Jaroslaw Kowalski nor the names of its 
+//   contributors may be used to endorse or promote products derived from this
+//   software without specific prior written permission. 
+// 
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE 
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE 
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE 
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR 
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS 
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN 
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) 
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF 
+// THE POSSIBILITY OF SUCH DAMAGE.
+// 
+
+namespace NLog.Targets
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Globalization;
+    using System.Text;
+    using NLog.Config;
+    using NLog.Internal;
+    using NLog.Layouts;
+
+    /// <summary>
+    /// Represents target that supports context capture using MDLC, MDC, NDLC and NDC
+    /// </summary>
+    public abstract class TargetWithContext : TargetWithLayout, IIncludeContext
+    {
+        /// <summary>Internal Layout that allows capture of MDC context</summary>
+        public Layout InternalLayoutMdc { get; }
+        /// <summary>Internal Layout that allows capture of NDC context</summary>
+        public Layout InternalLayoutNdc { get; }
+#if !SILVERLIGHT
+        /// <summary>Internal Layout that allows capture of MDLC context</summary>
+        public Layout InternalLayoutMdlc { get; }
+        /// <summary>Internal Layout that allows capture of NDLC context</summary>
+        public Layout InternalLayoutNdlc { get; }
+#endif
+        /// <summary>Internal Layout that allows capture of callsite context</summary>
+        public Layout InternalLayoutCallSite { get; }
```

Comment:
I think this is a "trick" for the objectScanner, is that correct?

I'm not sure if exposing those to public (guess it's needed now), is a good idea. I found it confusing and when I subclass this class, it won't understand what to do with those properties.

Reply:
if so, then my proposal

- make these layouts private
- make a new interface: 

```c#
internal interface IInternalLayouts{ //or maybe public
Layout[] getLayouts(); //or list/or ienumerable
}
```

- implement this interface explicitly (so members are hidden).
- check first for `IInternalLayouts`, combine them with the other found Layouts

Reply:
I have found a better trick, that I'm working on now. I will create nested Layouts, where the TopLevel becomes the Target.Layout (And will be scanned). The TopLevel will be similar to CompoundLayout.

Then override the Target.Layout-getter/setter so it will just be a single sub-Layout that lives under the TopLevel-Layout (along with all the context layouts).

Reply:
Smart one!